src/worldgen/DensityFunction.ts

```diff
@@ -307,21 +325,21 @@ export namespace DensityFunction {
 		public compute({ x: blockX, y: blockY, z: blockZ }: DensityFunction.Context) {
 			const w = this.cellWidth
 			const h = this.cellHeight
-			const x = (blockX % w) / w
-			const y = (blockY % h) / h
-			const z = (blockZ % w) / w
+			const x = ((blockX % w + w) % w) / w
+			const y = ((blockY % h + h) % h) / h
+			const z = ((blockZ % w + w) % w) / w
 			const firstX = Math.floor(blockX / w) * w
 			const firstY = Math.floor(blockY / h) * h
 			const firstZ = Math.floor(blockZ / w) * w
-			const noise000 = this.computeCorner(firstX, firstY, firstZ)
-			const noise001 = this.computeCorner(firstX, firstY, firstZ + w)
-			const noise010 = this.computeCorner(firstX, firstY + h, firstZ)
-			const noise011 = this.computeCorner(firstX, firstY + h, firstZ + w)
-			const noise100 = this.computeCorner(firstX + w, firstY, firstZ)
-			const noise101 = this.computeCorner(firstX + w, firstY, firstZ + w)
-			const noise110 = this.computeCorner(firstX + w, firstY + h, firstZ)
-			const noise111 = this.computeCorner(firstX + w, firstY + h, firstZ + w)
-			return lerp3(x, y, z, noise000, noise100, noise010, noise110, noise001, noise101, noise011, noise111)
+			const noise000 = () => this.computeCorner(firstX, firstY, firstZ)
+			const noise001 = () => this.computeCorner(firstX, firstY, firstZ + w)
+			const noise010 = () => this.computeCorner(firstX, firstY + h, firstZ)
+			const noise011 = () => this.computeCorner(firstX, firstY + h, firstZ + w)
+			const noise100 = () => this.computeCorner(firstX + w, firstY, firstZ)
+			const noise101 = () => this.computeCorner(firstX + w, firstY, firstZ + w)
+			const noise110 = () => this.computeCorner(firstX + w, firstY + h, firstZ)
+			const noise111 = () => this.computeCorner(firstX + w, firstY + h, firstZ + w)
+			return lazyLerp3(x, y, z, noise000, noise100, noise010, noise110, noise001, noise101, noise011, noise111)
```

Comment:
especially when only calculating individual slices, we don't need to calculate 8 values for each position, 4 are enough, so this lazyLerp doesn't calculate the corners it doesn't need.